components/speedreader/rust/lib/src/readability/src/extractor.rs

```diff
@@ -203,6 +416,137 @@ pub fn clean_title(dom: &Sink, title: String) -> String {
     }
 }
 
+#[derive(Error, Debug, PartialEq)]
+pub enum JsonLdError {
+    #[error("Could not parse json-ld: `{0}`")]
+    ParseError(String),
+    #[error("Missing @context field")]
+    MissingContext,
+    #[error("Invalid @context: Does not match http://schema.org")]
+    InvalidContext,
+    #[error("Missing @type field")]
+    MissingType,
+    #[error("Invalid @type")]
+    InvalidType,
+}
+
+impl From<serde_json::Error> for JsonLdError {
+    fn from(err: serde_json::Error) -> Self {
+        JsonLdError::ParseError(err.to_string())
+    }
+}
+
+#[inline(never)]
+fn try_parse_author(v: &serde_json::Value) -> Option<String> {
+    match v {
+        serde_json::Value::String(s) => {
+            if s.starts_with("[{") {
+                // Looks like it's nested JSON. I have no idea why sites do this. It's so stupid, but some do...
+                let p = serde_json::from_str(s).ok()?;
+                try_parse_author(&p)
+            } else {
+                Some(s.to_string())
+            }
+        }
+        serde_json::Value::Array(a) => {
+            if !a.is_empty() {
+                Some(
+                    a.iter()
+                        .filter_map(|e| try_parse_author(e))
+                        .collect::<Vec<_>>()
+                        .join(", "),
+                )
+            } else {
+                None
+            }
+        }
+        serde_json::Value::Object(o) => o
+            .get("name")
+            .and_then(|name| name.as_str())
```

Comment:
this seems like one more str conversation than is needed.  Why the as_str and to_string as seperate steps?  Could it `and_then` step just convert to str directly?

Reply:
ah, so the naming here is pretty confusig.

Basically we have no idea what the json structure is for these JSON-LD blobs. So when we call `o.get("name")` it will return us an `Option<serde_json::Value>`. We use the `and_then` to convert it to `Option<&str>`, then the final map copies out the resulting string from the json struct.